validator/server_arb/execution_run_test.go

```diff
@@ -0,0 +1,208 @@
+package server_arb
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/offchainlabs/nitro/validator"
+)
+
+type mockMachine struct {
+	gs         validator.GoGlobalState
+	totalSteps uint64
+}
+
+func (m *mockMachine) Hash() common.Hash {
+	if m.gs.PosInBatch == m.totalSteps-1 {
+		return machineFinishedHash(m.gs)
+	}
+	return m.gs.Hash()
+}
+
+func (m *mockMachine) GetGlobalState() validator.GoGlobalState {
+	return m.gs
+}
+
+func (m *mockMachine) Step(ctx context.Context, stepSize uint64) error {
+	for i := uint64(0); i < stepSize; i++ {
+		if m.gs.PosInBatch == m.totalSteps-1 {
+			return nil
+		}
+		m.gs.PosInBatch += 1
+	}
+	return nil
+}
+
+func (m *mockMachine) CloneMachineInterface() MachineInterface {
+	return &mockMachine{
+		gs:         validator.GoGlobalState{Batch: m.gs.Batch, PosInBatch: m.gs.PosInBatch},
+		totalSteps: m.totalSteps,
+	}
+}
+func (m *mockMachine) GetStepCount() uint64 {
+	return 0
+}
+func (m *mockMachine) IsRunning() bool {
+	return m.gs.PosInBatch < m.totalSteps-1
+}
+func (m *mockMachine) ValidForStep(uint64) bool {
+	return true
+}
+func (m *mockMachine) Status() uint8 {
+	if m.gs.PosInBatch == m.totalSteps-1 {
+		return uint8(validator.MachineStatusFinished)
+	}
+	return uint8(validator.MachineStatusRunning)
+}
+func (m *mockMachine) ProveNextStep() []byte {
+	return nil
+}
+func (m *mockMachine) Freeze()  {}
+func (m *mockMachine) Destroy() {}
+
+func Test_machineHashesWithStep(t *testing.T) {
+	mm := &mockMachine{}
+	e := &executionRun{}
+	ctx := context.Background()
+
+	t.Run("basic argument checks", func(t *testing.T) {
```

Comment:
Why are all these different test cases being run using t.Run instead of just being separate top-level test functions?

Reply:
Personal preference here. I find it more expressive to write subcases in plain english instead of having to wrangle the Go function naming convention. It also groups functionality into one place, but I'm happy to change it if the alternative is highly preferred